nucliadb/nucliadb/train/generators/paragraph_streaming.py

```diff
@@ -0,0 +1,93 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+from typing import AsyncIterator
+
+from nucliadb_protos.dataset_pb2 import (
+    ParagraphStreamingBatch,
+    ParagraphStreamItem,
+    TrainSet,
+)
+from nucliadb_protos.nodereader_pb2 import StreamRequest
+
+from nucliadb.common.cluster.base import AbstractIndexNode
+from nucliadb.ingest.orm.resource import KB_REVERSE
+from nucliadb.train import logger
+from nucliadb.train.generators.utils import get_resource_from_cache_or_db
+
+
+async def generate_paragraph_streaming_payloads(
+    kbid: str,
+    trainset: TrainSet,
+    node: AbstractIndexNode,
+    shard_replica_id: str,
+) -> AsyncIterator[ParagraphStreamingBatch]:
+    """Streams paragraphs ordered as if they were read sequentially from each
+    field.
+
+    """
+    request = StreamRequest()
+    request.shard_id.id = shard_replica_id
+
+    batch = ParagraphStreamingBatch()
+    async for document_item in node.stream_get_fields(request):
+        field_id = f"{document_item.uuid}{document_item.field}"
+        rid, field_type, field = field_id.split("/")
+
+        orm_resource = await get_resource_from_cache_or_db(kbid, rid)
+        if orm_resource is None:
+            logger.error(f"{rid} does not exist on DB")
+            continue
+
+        field_type_int = KB_REVERSE[field_type]
+        field_obj = await orm_resource.get_field(field, field_type_int, load=False)
+
+        extracted_text = await field_obj.get_extracted_text()
+        field_metadata = await field_obj.get_field_metadata()
+        if field_metadata is None:
+            logger.error(f"{field_id} does not have field metadata on DB")
+            continue
+
+        for paragraph in field_metadata.metadata.paragraphs:
+            item = ParagraphStreamItem()
+            item.id = f"{rid}/{field_type}/{field}/{paragraph.start}-{paragraph.end}"
+            item.text = extracted_text.text[paragraph.start : paragraph.end]
+
+            batch.data.append(item)
+            if len(batch.data) == trainset.batch_size:
+                yield batch
+                batch = ParagraphStreamingBatch()
+
+        for split, metadata in field_metadata.split_metadata.items():
+            # REVIEW: do we have to care about ExtractedText.deleted_splits?
+            split_text = extracted_text.split_text[split]
+
+            for paragraph in metadata.paragraphs:
+                item = ParagraphStreamItem()
+                item.id = f"{rid}/{field_type}/{field}/{split}/{paragraph.start}-{paragraph.end}"
+                item.text = split_text[paragraph.start : paragraph.end]
+
+                batch.data.append(item)
+                if len(batch.data) == trainset.batch_size:
+                    yield batch
+                    batch = ParagraphStreamingBatch()
+
+    if len(batch.data):
+        yield batch
```

Comment:
nit: you could add the batching logic into a separate function. This would ease testing too.

Reply:
Do you mean a generic "batchify" function that iterates an iterator and return batches? I'll do it for all generators then

Reply:
yes, exactly. Something that is able to batch in a generic way